operator/main.go

```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
 	}
 	close(k8sInitDone)
 
+	k8sversion.Update(k8s.Client(), option.Config)
+	if !k8sversion.Capabilities().MinimalVersionMet {
+		log.Fatalf("Minimal kubernetes version not met: %s < %s",
+			k8sversion.Version(), k8sversion.MinimalVersionConstraint)
+	}
+
+	// Get hostname for identity name of the lease lock holder.
+	// We identify the leader of the operator cluster using hostname.
+	operatorID, err := os.Hostname()
+	if err != nil {
+		log.WithError(err).Fatal("failed to get hostname")
+	}
+	operatorID = rand.RandomStringWithPrefix(operatorID+"-", 10)
+
+	ns := option.Config.K8sNamespace
+	// If due to any reason the NAMESPACE is not set we assume it to be
+	// in default namespace.
+	if ns == "" {
+		ns = metav1.NamespaceDefault
+	}
+
+	leaseLock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      leaderElectionLeaseLockName,
+			Namespace: ns,
+		},
+		Client: k8s.Client().CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			// Identity name of the holder
+			Identity: operatorID,
+		},
+	}
+
+	// Start the leader election for running cilium-operators
+	leaderelection.RunOrDie(leaderElectionCtx, leaderelection.LeaderElectionConfig{
+		Name: leaderElectionLeaseLockName,
+
+		Lock:            leaseLock,
+		ReleaseOnCancel: true,
+
+		LeaseDuration: operatorOption.Config.LeaderElectionLeaseDuration,
+		RenewDeadline: operatorOption.Config.LeaderElectionRenewDeadline,
+		RetryPeriod:   operatorOption.Config.LeaderElectionRetryPeriod,
+
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: onOperatorStartLeading,
+			OnStoppedLeading: func() {
+				log.WithField("operator_id", operatorID).Info("leader election lost")
+				// Cleanup everything here, and exit.
+				gops.Close()
+				close(shutdownSignal)
+				leaderElectionCtxCancel()
+				os.Exit(0)
```

Comment:
More of a question here: is the `Exit` in order for the previous cilium-operator leader to become a "follower" (non-leader)?

Reply:
I don't think if I understand the question correctly. :sweat_smile: 

The reason why we do an exit here is that once an operator stops leading, there is a lot of complexity in terms of stopping all the watchers gracefully managed by the operator. A more general approach I have seen with leader election in operators is to exit when once an already leading instance stops leading. Like here in tidb-operator - https://github.com/pingcap/tidb-operator/blob/master/cmd/controller-manager/main.go#L237-L239

Reply:
I took a look at what kcm/scheduler do and both when they stop leading call klog.Fatalf and exit. 


I suspect this is going to cause some issues for the operator at least in the AWS ENI mode.

Think of the following case:
- user runs operator and configures the CiliumNode to set DeleteOnTermination for the ENIs that get created
- CN comes online and operator creates the ENI in AWS
- operator loses the leadership but it did not get a change to attach the ENI and set delete on termination
- user is left with a dangling ENI in the AWS account

I don't think we should over-engineer a solution now but should be aware of these edge cases in the future.

Reply:
That's a good point. Cilium can also crash in the middle of that operation. Is it possible to know which ENIs are dangling? I'm just thinking on the case where we add a GC to the operator to clean up such dangling ENIs.

Reply:
Yep -- we can do it based on the tags (or names) of the ENIs. I think this exposes the fact that the operator is doing a lot of things that are not atomic and we do not have a good way to cancel/repair those operations when we lose leadership.

Reply:
This is a good discussion and I've opened an issue to track these things. Please add whatever comes to mind in issue https://github.com/cilium/cilium/issues/12448

Reply:
Essentially I was asking if the `Exit` is for transitioning between leader to non-leader, and it looks like the answer is yes. @fristonio